fvm/src/call_manager/default.rs

```diff
@@ -285,57 +277,113 @@ where
         let mut linker = Linker::new(&engine);
         bind_syscalls(&mut linker).or_fatal()?;
 
+        log::trace!("calling {} -> {}::{}", from, to, method);
         self.map_mut(|cm| {
             // Make the kernel/store.
             let kernel = K::new(cm, from, to, method, value.clone());
-            let mut store = Store::new(&engine, kernel);
-
-            log::trace!("calling {} -> {}::{}", from, to, method);
-
-            let result = (|| {
-                // Load parameters, if there are any.
-                let param_id = if params.len() > 0 {
-                    store.data_mut().block_create(DAG_CBOR, params)?
-                } else {
-                    super::NO_DATA_BLOCK_ID
-                };
+            let mut store = Store::new(&engine, InvocationData::new(kernel));
+
+            // Store parameters, if any.
+            let param_id = if params.len() > 0 {
+                match store.data_mut().kernel.block_create(DAG_CBOR, params) {
+                    Ok(id) => id,
+                    // This could fail if we pass some global memory limit.
+                    Err(err) => return (Err(err), store.into_data().kernel.take()),
+                }
+            } else {
+                super::NO_DATA_BLOCK_ID
+            };
 
+            // From this point on, there are no more syscall errors, only aborts.
+            let result: std::result::Result<RawBytes, Abort> = (|| {
                 // Instantiate the module.
-                let instance = linker.instantiate(&mut store, &module).or_fatal()?;
+                let instance = linker
+                    .instantiate(&mut store, &module)
+                    // The module should already have been checked by the system.
+                    .map_err(Abort::Fatal)?;
+
+                // Lookup the invoke method.
+                let invoke: wasmtime::TypedFunc<(u32,), u32> = instance
```

Comment:
I just moved the type up here and made it explicit because I get nervous about these things.

Reply:
Paranoia pays off ;-)